binary_search_tree/tree.py

```diff
@@ -14,44 +14,157 @@ class Tree:
+                return None
+            else:
+                return self.helper_find(key, current.left)
+        else:
+            if current.right == None:
+                return None
+            else:
+                return self.helper_find(key, current.right)
+
 
-    # Time Complexity: 
-    # Space Complexity: 
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def inorder(self):
-        pass
+        result = []
+        self.helper_inorder(self.root, result)
+        return result
 
-    # Time Complexity: 
-    # Space Complexity:     
+    def helper_inorder(self, current, result):
+        if current == None:
+            return
+        elif current.left == None and current.right == None:
+            result.append({"key": current.key, "value": current.value})
+        else:
+            self.helper_inorder(current.left, result)
+            result.append({"key": current.key, "value": current.value})
+            self.helper_inorder(current.right, result)
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)  
     def preorder(self):
-        pass
+        result = []
+        self.helper_preorder(self.root, result)
+        return result
+    
+    def helper_preorder(self, current, result):
+        if current == None:
+            return
+        elif current.left == None and current.right == None:
+            result.append({"key": current.key, "value": current.value})
+        else:
+            result.append({"key": current.key, "value": current.value})
+            self.helper_preorder(current.left, result)
+            self.helper_preorder(current.right, result)
+    
+    def preorder2(self):
+        result = []
+        if not self.root:
+            return result
+        stack = [self.root]
+        while len(stack) != 0:
+            current = stack.pop()
+            result.append({"key": current.key, "value": current.value})
+            if current.right:
+                stack.append(current.right)
+            if current.left:
+                stack.append(current.left)
+
+        return result
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(n) 
     def postorder(self):
-        pass
+        result = []
+        self.helper_postorder(self.root, result)
+        return result
 
-    # Time Complexity: 
-    # Space Complexity:     
+    def helper_postorder(self, current, result):
+        if current == None:
+            return
+        elif current.left == None and current.right == None:
+            result.append({"key": current.key, "value": current.value})
+        else:
+            self.helper_postorder(current.left, result)
+            self.helper_postorder(current.right, result)
+            result.append({"key": current.key, "value": current.value})
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n) because there are n function calls on the stack? 
     def height(self):
```

Comment:
✨Nice! Time and space complexity would be O(n) if the tree is unbalanced! For a balanced tree, time and space complexity would be O(log n) because your recursive calls only go through half the nodes!